File: dcp-solutions/2025-01-21/solution.c
```c
#include <limits.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    struct TreeNode *parent;
};

int inorderSuccessor(struct TreeNode* root, struct TreeNode* node) {
    if (node->right != NULL && searchNode(node->right, node->val)) return findNextBigger(root, node);
    else {
        while (node->parent != NULL && node == findLeftMost(node->parent, node->val))
            node = node->parent;
        return findNextBigger(root, node->parent);
    }
}

int searchNode(struct TreeNode* root, int val) {
    if (root == NULL || root->val >= val)
        return 0;
    else
        return searchNode(root->left, val) + searchNode(root->right, val) + 1;
}

struct TreeNode* findLeftMost(struct TreeNode* node, int val) {
    while (node != NULL && node->val <= val)
        node = node->right;
    return node;
}

int findNextBigger(struct TreeNode* root, struct TreeNode* node) {
    if (node == root) return INT_MAX;

    if (root->left != NULL && searchNode(root->left, node->val)) 
        return findLeftMost(root->parent, node->val);

    if (root->right != NULL && searchNode(root->right, node->val))
        return findLeftMost(root, node->val);
    
    struct TreeNode* successor = root->right;
    while(successor != NULL && successor->left == node)
        successor = successor->right;

    int temp = node->val;
    node->val = successor->val;
    successor->val = temp;

    return findLeftMost(root, node->val);
}
```

File: dcp-solutions/2025-01-21/solution.php
```php
class TreeNode {
    public $val;
    public $left;
    public $right;
    public $parent;

    function __construct($val) {
        $this->val = $val;
        $this->left = null;
        $this->right = null;
        $this->parent = null;
    }
}

class Solution {
    function inorderSuccessor($root, $p) {
        if (!$p || !$root || $p == $root) return null;

        while ($p && $p->right && $p->right != $root) {
            $p = $p->right;
        }

        if ($p->right == $root) {
            $successor = $this->dfs($root, $p);
            return $successor ? $successor->val : null;
        } else {
            while ($p && $p->left != $root) {
                $p = $p->left;
            }
            return $this->dfs($root, $p);
        }
    }

    function dfs($node, $p) {
        if (!$node || $node == $p) return null;

        while ($node && $node->right && $node->right != $p) {
            $node = $node->right;
        }

        if ($node->right == $p) {
            return $this->dfs($node, $p);
        } else {
            while ($node && $node->left != $p) {
                $node = $node->left;
            }
            return $this->dfs($node, $p);
        }
    }
}
```